README.md

```diff
@@ -99,20 +86,56 @@ Note that you may not get cached results if you use a different `nix` version th
+
+### Structure of `survey/default.nix`
+
+The process of building up to the final `haskellPackages` is broken down into multiple steps, with the intermediate package sets also exposed for your use.
+
+We start with a plain nixpkgs named `normalPkgs`, defaulting to the version provided by [`nixpkgs.nix`](./nixpkgs.nix), but you can pass in your own version here.
+You could instead provide `overlays`, which get applied to `normalPkgs`.
+The `.pkgsMusl` from `normalPkgs` now forms our base `pkgs`.
+
+Packages in `pkgsMusl` typically [only include `.so` files](https://github.com/NixOS/nixpkgs/issues/61575), but not `.a` files. We create a `archiveFilesOverlay`, which overrides our Haskell dependencies (i.e. C libraries) to include `.a` files as well. 
+This usually involves low-level library specific actions, but the goal is to upstream these to nixpkgs under a `dontDisableStatic` attribute.
+We apply this overlay to `pkgs` to get `pkgsWithArchiveFiles`.
+
+This package set _may_ be sufficient for your needs if you have a separate build system (e.g. Bazel) and are only after a statically linked GHC and library dependencies.
+
+The next step is to configure Hackage library packages to use static linking. This is done by the `haskellLibsReadyForStaticLinkingOverlay` to produce `pkgsWithHaskellLibsReadyForStaticLinking`.
+TODO: Document what `addStaticLinkerFlagsWithPkgconfig` does.
+
+Finally `staticHaskellBinariesOverlay` infers which haskell packages are binaries, and overrides them using the `statify` function, which primarily sets the `--enable-executable-static` Cabal flag.
+This produces `pkgsWithStaticHaskellBinaries`, which is where the `haskellPackages` attribute set is taken from.
+
+### Building existing packages in the survey
+
+To build existing packages, run:
 
 ```
 NIX_PATH=nixpkgs=nixpkgs nix-build --no-link survey/default.nix -A working
 ```
 
-There are multiple package sets available in the survey (select via `-A`):
+Relevant package sets available in the survey include (select via `-A`):
 
 * `working` -- build all exes known to be working
 * `notWorking` -- build all exes known to be not working (help welcome to make them work)
 * `haskellPackages.somePackage` -- build a specific package from our overridden package set
 
-If you are a nix user, you can easily `import` this functionality and add an override to add your own packages.
+## Building a minimal example (don't use this in practice)
+
+`default.nix` builds an example executable (originally from https://github.com/vaibhavsagar/experiments). Run:
+
+```
+NIX_PATH=nixpkgs=nixpkgs nix-build --no-link
+```
+
+This prints a path that contains the fully linked static executable in the `bin` subdirectory.
+
+This example is so that you get the general idea.
+In practice, you probably want to use one of the approaches from the "Building arbitrary packages" or "Building stack projects" sections below.
```

Comment:
I would probably still keep that at the top, so that the user gets a rough idea of what's happening by looking at a short file which isn't thousands of lines long (we might state that even more explicitly, and remove the `probably`).

If we do move it down, then `sections below` needs to be updated.

Reply:
Ok, that's fair — my motivation was that we should tell readers _how_ to do something first, before telling them _how not_ to do it. 

I think we should replace the minimal example entirely with instructions on how to use the survey, perhaps like 
https://github.com/nh2/static-haskell-nix/issues/88#issuecomment-1646615103, but this is out of scope of this PR and can be discussed later.